### PR TITLE
Add Support for Metadata on Enumerators

### PR DESCRIPTION
**Pull request by InsertCreativityHere**
This PR adds support for metadata on enumerators, and simplifies the code for constructing enumerators (#2515):

----

Currently, `createEnumerator` and `validateEnumerator` are on `Container`, not `Enum`.
And `Enum` has it's own function named `newEnumerator`.

All 3 of these functions were simplified and merged into a single `createEnumerator` on `Enum` (like you'd expect).

### CODE REVIEW

File: cpp/src/Slice/Grammar.y [hunk 1]
```diff
@@ -1674,41 +1689,33 @@ enumerator
 : ICE_IDENTIFIER
```
Comment (InsertCreativityHere) [inline cpp/src/Slice/Grammar.y:1689]: Also changed the parsing rule for `enumerator`.
For some reason, it's currently returning an `EnumeratorList` with one element in it...
I updated the code to just return a normal `EnumeratPtr` instead.

File: cpp/src/Slice/Grammar.cpp
Comment (InsertCreativityHere) [inline cpp/src/Slice/Grammar.cpp:1]: This is all generated code, no need to look at it unless you really want to.

File: cpp/test/Slice/errorDetection/EnumeratorValues.err [hunk 2]
```diff
@@ -1,5 +1,5 @@
 EnumeratorValues.ice:12: value for enumerator `C' is out of range
-EnumeratorValues.ice:12: enumerator `C' has the same value as enumerator `B'
 EnumeratorValues.ice:13: value for enumerator `D' is out of range
 EnumeratorValues.ice:14: value for enumerator `E' is out of range
 EnumeratorValues.ice:15: value for enumerator `F' is out of range
+EnumeratorValues.ice:15: enumerator `F' has the same value as enumerator `B'
```
Comment (InsertCreativityHere) [inline cpp/test/Slice/errorDetection/EnumeratorValues.err:5]: My changes have a small impact on the ordering/logic we use to check for duplicate enumerators,
hence this small change in the error detection test.

**Review by externl — APPROVED** [review]

File: cpp/src/Slice/Grammar.y [hunk 3]
```diff
@@ -1674,41 +1689,33 @@ enumerator
...
-    $$ = ens;
+    EnumPtr cont = dynamic_pointer_cast<Enum>(currentUnit->currentContainer());
+    $$ = cont->createEnumerator(ident->v, nullopt);
 }
 | ICE_IDENTIFIER '=' enumerator_initializer
 {
     auto ident = dynamic_pointer_cast<StringTok>($1);
-    auto ens = make_shared<EnumeratorListTok>();
-    ContainerPtr cont = currentUnit->currentContainer();
+    EnumPtr cont = dynamic_pointer_cast<Enum>(currentUnit->currentContainer());
```
Comment (bernardnormier) [inline cpp/src/Slice/Grammar.y:1698]: You don't check that cont is not null because an enumerator can only be defined in an Enum container?

File: cpp/src/Slice/Grammar.y [hunk 4]
```diff
@@ -1653,14 +1653,29 @@ ICE_ENUM
...
 {
+    auto metadata = dynamic_pointer_cast<StringListTok>($1);
+    auto enumerator = dynamic_pointer_cast<Enumerator>($2);
+    if (enumerator && !metadata->v.empty())
+    {
+        enumerator->setMetaData(metadata->v);
+    }
+    auto enumeratorList = make_shared<EnumeratorListTok>();
+    enumeratorList->v.push_front(enumerator);
+    $$ = enumeratorList;
```
Comment (bernardnormier) [inline cpp/src/Slice/Grammar.y:1678]: It's not totally clear to me why we have the %empty rule. Presumably the 'meta_data enumerator' rule above is sufficient to terminate the parsing of a non-empty enumeration.

According to:
https://doc.zeroc.com/ice/3.7/the-slice-language/user-defined-types/enumerations

> Slice does not permit empty enumerations.

Is this a bug and the parser actually allows empty enumerations?

In new Slice:
https://docs.icerpc.dev/slice2/language-guide/enum-types

> A checked enumeration must have at least one enumerator, while an unchecked enumeration may have no enumerator at all.

File: cpp/src/Slice/Grammar.y [hunk 5]
```diff
@@ -1674,41 +1689,33 @@ enumerator
...
-        else
-        {
-            EnumeratorPtr en = cont->createEnumerator(ident->v, static_cast<int>(intVal->v));
-            ens->v.push_front(en);
-        }
+        $$ = cont->createEnumerator(ident->v, static_cast<int>(intVal->v));
+    }
+    else
+    {
+        $$ = cont->createEnumerator(ident->v, nullopt); // Dummy
```
Comment (bernardnormier) [inline cpp/src/Slice/Grammar.y:1710]: I don't understand the dummy comment.

File: cpp/src/Slice/Grammar.y [hunk 6]
```diff
@@ -1674,41 +1689,33 @@ enumerator
...
     auto ident = dynamic_pointer_cast<StringTok>($1);
-    auto ens = make_shared<EnumeratorListTok>();
-    ContainerPtr cont = currentUnit->currentContainer();
+    EnumPtr cont = dynamic_pointer_cast<Enum>(currentUnit->currentContainer());
     auto intVal = dynamic_pointer_cast<IntegerTok>($3);
     if (intVal)
     {
         if (intVal->v < 0 || intVal->v > std::numeric_limits<int32_t>::max())
         {
             currentUnit->error("value for enumerator `" + ident->v + "' is out of range");
```
Comment (bernardnormier) [inline cpp/src/Slice/Grammar.y:1704]: What happens after you issue this error?

My understanding is the existing code tried to keep going.

File: cpp/src/Slice/Parser.h [hunk 7]
```diff
@@ -896,8 +893,7 @@ namespace Slice
     class Enumerator final : public virtual Contained
     {
     public:
-        Enumerator(const ContainerPtr&, const std::string&, std::optional<int>);
-        void init() final;
+        Enumerator(const ContainerPtr&, const std::string&, int, bool);
```
Comment (bernardnormier) [inline cpp/src/Slice/Parser.h:896]: Please give names to parameters.

Comment (bernardnormier) [reply cpp/src/Slice/Parser.h:896]: I also find the previous `std::optional<int>` nicer.

**Review by bernardnormier — APPROVED** [review]

File: cpp/src/Slice/Parser.h (hunk 7 above)
Comment (InsertCreativityHere) [reply cpp/src/Slice/Parser.h:896]: > Please give names to parameters.

Alright, I'll do that in a follow up PR.
None of the grammar headers use named parameters. I have no idea why, that's just how it always was.

> I also find the previous std::optional<int> nicer

It doesn't really work here though.
It's probably hard to follow without names, but take:
```
enum Foo { A }
```
Here the enumerator has a name `A`, a value of `0`, and `hasExplicitValue` (the bool) is `false`.
With an `optional`, there's no way to still provide a value in the case when it's '`false`' (ie. empty).

File: cpp/src/Slice/Grammar.y (hunk 6 above)
Comment (InsertCreativityHere) [reply cpp/src/Slice/Grammar.y:1704]: Yes, the new code also keeps going. On the line right under this we set:
`$$ = cont->createEnumerator(ident->v, static_cast<int>(intVal->v));`
this is effectively a return statement in Bison. So we still construct the enumerator, and return it.

File: cpp/src/Slice/Grammar.y (hunk 5 above)
Comment (InsertCreativityHere) [reply cpp/src/Slice/Grammar.y:1710]: This case only occurs when the cast `auto intVal = dynamic_pointer_cast<IntegerTok>($3)` fails.
But this grammar rule is only triggered when there's an integer literal present...

So something must of gone wrong for the cast to fail. To be honest, I didn't look too closely at this specific case,
but it might just be impossible to end up in this case after thinking about it now.

File: cpp/src/Slice/Grammar.y (hunk 4 above)
Comment (InsertCreativityHere) [reply cpp/src/Slice/Grammar.y:1678]: > Is this a bug 

nope, ...

> and the parser actually allows empty enumerations?

... and yes!

----

You're correct, empty enums are disallowed in Slice, but we still have the logic to parse them correctly.
If we didn't have this rule, then an empty enum would cause a `syntax error: expected identifier, '[',  ...` to be emitted:
A) this error message isn't super helpful
B) and it isn't completely accurate, an empty enum is a _semantic_ error, not a _syntax_ error.
C) The bison compiler has no syntax-error recovery, so we only emit _true_ syntax errors as a last resort.

So, the idea is, we allow this at the parser level, and the place where we actually use this rule is where we do the check.
This way we can gracefully handle the error, and provide a more helpful error message. See:
https://github.com/zeroc-ice/ice/blob/6980c561cd479a6e5794193fe3e80b1eafaced83/cpp/src/Slice/Grammar.y#L1623-L1636

File: cpp/src/Slice/Grammar.y (hunk 3 above)
Comment (InsertCreativityHere) [reply cpp/src/Slice/Grammar.y:1698]: Yes, by the time we hit this parsing rule, if there isn't a container, something has already gone horribly wrong.

File: cpp/src/Slice/Grammar.y (hunk 4 above)
Comment (InsertCreativityHere) [reply cpp/src/Slice/Grammar.y:1678]: Just as a note, if you ever look through the grammar you'll actually see alot of this.
'bugs' where the parser accepts completely bogus Slice!

For the reason that if we don't, we'll get Bison's default error message, and parsing halts immediately; both not great.
So, we allow the bogus Slice _for a little bit_ until we hit a place where we can safely reject it with a custom error message.

File: cpp/src/Slice/Parser.h (hunk 7 above)
Comment (bernardnormier) [reply cpp/src/Slice/Parser.h:896]: ok

Comment (InsertCreativityHere) [reply cpp/src/Slice/Parser.h:896]: IMO it's just more straightforward this way. `Enumerator` has always held it's actual value (explicit or implicit), since often, that's all the code gen cares about. A key fact is that `Enumerator` on it's own cannot compute an implicit value for itself.
Only `Enum` has enough information to compute that.

The old logic was `Enum::createEnumerator` would set a special value of `-1` if none was provided.
And the constructor `Enumerator::Enumerator` would call back into `Enum::newEnumerator` to check for this special value and patch it with a real value.

But... there's no benefit to delaying this, instead of just computing the value up front. `Enum` already has all the necessary information. No need to jump back and forth between `Enum` and `Enumerator`, and this avoids storing a special 'replace me' value of `-1` too.
